File: src/js/lib/rx.js
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

import { AsciiCharSet } from './ascii-char-set.js';

/**
 * What an {@link Rx} node is.
 *
 * @enum {string}
 */
export const RxKind = Object.freeze({
	/** The empty language, which arises only as a derivative. */
	EmptySet: 'EmptySet',
	/** The language holding only the empty string. */
	Epsilon: 'Epsilon',
	/** A non-empty set of characters matching one position. */
	Chars: 'Chars',
	/** Two or more expressions in sequence. */
	Concat: 'Concat',
	/** Two or more expressions in alternation. */
	Union: 'Union',
	/** An expression repeated between `minCount` and `maxCount` times. */
	Interval: 'Interval',
});

/**
 * What `maxLength` saturates at.
 *
 * The C# saturates at `long.MaxValue`; this is smaller, and the difference cannot show. A
 * language whose longest string is *n* needs at least *n* + 1 states, so anything whose length
 * approaches this is refused by the state budget long before the ordering that reads `maxLength`
 * is reached. Saturation only keeps the arithmetic honest while converting a naxp that is going
 * to be refused anyway.
 */
const MAX_LENGTH = Number.MAX_SAFE_INTEGER;

const NO_CHILDREN = Object.freeze([]);

/**
 * An expression in the algebra the state map is built over.
 *
 * This is deliberately not an `Ast`. The tree records what was written, whereas these nodes are
 * what derivatives are taken of: they carry an empty language, they are normalised by their
 * factory, and they are interned, so equal expressions are the same object and identity can be
 * relied on as a map key.
 *
 * Normalisation does not have to reduce every expression denoting the same language to one form,
 * and it does not. Making the machine canonical is the job of hash-consing on transition lists in
 * `StateMapBuilder`; normalisation here only keeps derivatives from growing and makes memoisation
 * bite.
 *
 * Intervals stay symbolic. Expanding `(A{99}){99}` into nearly ten thousand nodes would throw
 * away the reason the count cap exists.
 */
export class Rx {
	/**
	 * @param {number} id A number unique within the factory that made this node.
	 * @param {string} kind One of {@link RxKind}.
	 * @param {AsciiCharSet} charSet The characters, for `Chars`.
	 * @param {Rx[]} children The operands, for `Concat`, `Union` and `Interval`.
	 * @param {number} minCount The fewest repetitions, for `Interval`.
	 * @param {number} maxCount The most repetitions, for `Interval`.
	 * @param {boolean} isNullable Whether the language holds the empty string.
	 * @param {number} maxLength The length of the longest string in the language.
	 */
	constructor(id, kind, charSet, children, minCount, maxCount, isNullable, maxLength) {
		this.id = id;
		this.kind = kind;
		this.charSet = charSet;
		this.children = children;
		this.minCount = minCount;
		this.maxCount = maxCount;
		this.isNullable = isNullable;

		/**
		 * The length of the longest string in the language, or zero where the language is empty.
		 *
		 * Exact rather than an upper bound, and it strictly decreases along every derivative,
		 * which is what lets the builder order the states without a topological sort.
		 */
		this.maxLength = maxLength;

		/** @type {AsciiCharSet[] | null} */
		this.cachedFirstSets = null;
	}

	/**
	 * The character sets that can match the first character of a string in this language.
	 *
	 * These overlap in general. The minterms of them refine the first classes the specification
	 * defines, and the builder recovers the classes themselves by merging afterwards.
	 *
	 * @returns {AsciiCharSet[]} The sets.
	 */
	getFirstSets() {
		if (this.cachedFirstSets !== null) { return this.cachedFirstSets; }

		const sets = [];

		this.collectFirstSets(sets);
		this.cachedFirstSets = sets;

		return sets;
	}

	/**
	 * @param {AsciiCharSet[]} sets Where to put them.
	 */
	collectFirstSets(sets) {
		switch (this.kind) {
			case RxKind.EmptySet:
			case RxKind.Epsilon:
				return;

			case RxKind.Chars:
				sets.push(this.charSet);
				return;

			case RxKind.Concat:
				for (const child of this.children) {
					child.collectFirstSets(sets);

					if (!child.isNullable) { return; }
				}

				return;

			case RxKind.Union:
				for (const child of this.children) { child.collectFirstSets(sets); }
				return;

			case RxKind.Interval:
				this.children[0].collectFirstSets(sets);
				return;

			default:
				throw new Error(`Unhandled kind ${this.kind}.`);
		}
	}
}

/**
 * Makes {@link Rx} nodes, normalising and interning as it goes.
 *
 * One factory per build. Interning is not shared between naxps, so nothing accumulates.
 */
export class RxFactory {
	constructor() {
		/** @type {Map<string, Rx>} */
		this.interned = new Map();
		/** @type {Map<string, Rx>} */
		this.derivatives = new Map();
		this.nextId = 0;

		/** The empty language. */
		this.emptySet = this.intern(RxKind.EmptySet, AsciiCharSet.empty, NO_CHILDREN, 0, 0, false, 0);

		/** The language holding only the empty string. */
		this.epsilon = this.intern(RxKind.Epsilon, AsciiCharSet.empty, NO_CHILDREN, 0, 0, true, 0);
	}

	/** How many distinct expressions this factory has made. */
	get count() {
		return this.interned.size;
	}

	/**
	 * @param {AsciiCharSet} set The characters.
	 * @returns {Rx} The expression.
	 */
	chars(set) {
		return set.isEmpty
			? this.emptySet
			: this.intern(RxKind.Chars, set, NO_CHILDREN, 0, 0, false, 1);
	}

	/**
	 * Concatenation, flattened, with the empty string dropped and the empty language absorbing.
	 *
	 * @param {Rx[]} parts The operands, in order.
	 * @returns {Rx} The expression.
	 */
	concat(parts) {
		const flattened = [];

		for (const part of parts) {
			if (part.kind === RxKind.EmptySet) { return this.emptySet; }
			if (part.kind === RxKind.Epsilon) { continue; }

			if (part.kind === RxKind.Concat) { flattened.push(...part.children); }
			else { flattened.push(part); }
		}

		if (flattened.length === 0) { return this.epsilon; }
		if (flattened.length === 1) { return flattened[0]; }

		let isNullable = true;
		let maxLength = 0;

		for (const part of flattened) {
			isNullable = isNullable && part.isNullable;
			maxLength = saturatingAdd(maxLength, part.maxLength);
		}

		return this.intern(RxKind.Concat, AsciiCharSet.empty, flattened, 0, 0, isNullable, maxLength);
	}

	/**
	 * @param {Rx} first The first operand.
	 * @param {Rx} second The second operand.
	 * @returns {Rx} The concatenation.
	 */
	concatTwo(first, second) {
		return this.concat([first, second]);
	}

	/**
	 * Alternation, flattened, with the empty language dropped and duplicates removed.
	 *
	 * The operands are sorted by id. Ids differ between runs, but within one run two unions over
	 * the same operands sort the same way, which is all interning needs.
	 *
	 * @param {Rx[]} alternatives The operands.
	 * @returns {Rx} The expression.
	 */
	union(alternatives) {
		const flattened = [];

		for (const alternative of alternatives) {
			if (alternative.kind === RxKind.EmptySet) { continue; }

			if (alternative.kind === RxKind.Union) { flattened.push(...alternative.children); }
			else { flattened.push(alternative); }
		}

		flattened.sort((left, right) => left.id - right.id);

		const distinct = [];

		for (const alternative of flattened) {
			if (distinct.length === 0 || distinct[distinct.length - 1] !== alternative) {
				distinct.push(alternative);
			}
		}

		if (distinct.length === 0) { return this.emptySet; }
		if (distinct.length === 1) { return distinct[0]; }

		let isNullable = false;
		let maxLength = 0;

		for (const alternative of distinct) {
			isNullable = isNullable || alternative.isNullable;
			maxLength = Math.max(maxLength, alternative.maxLength);
		}

		return this.intern(RxKind.Union, AsciiCharSet.empty, distinct, 0, 0, isNullable, maxLength);
	}

	/**
	 * @param {Rx} first The first alternative.
	 * @param {Rx} second The second alternative.
	 * @returns {Rx} The alternation.
	 */
	unionTwo(first, second) {
		return this.union([first, second]);
	}

	/**
	 * Between `minCount` and `maxCount` copies in sequence.
	 *
	 * @param {Rx} child The expression repeated.
	 * @param {number} minCount The fewest repetitions.
	 * @param {number} maxCount The most repetitions.
	 * @returns {Rx} The expression.
	 */
	interval(child, minCount, maxCount) {
		if (maxCount === 0) { return this.epsilon; }
		if (child.kind === RxKind.Epsilon) { return this.epsilon; }
		if (child.kind === RxKind.EmptySet) { return minCount === 0 ? this.epsilon : this.emptySet; }

		// Where the child accepts the empty string, so does every count above the minimum, and
		// x{m,n} and x{0,n} are the same language. Normalising here is what lets isNullable be
		// read off the minimum alone.
		const low = child.isNullable ? 0 : minCount;

		if (low === 1 && maxCount === 1) { return child; }

		const maxLength = saturatingMultiply(child.maxLength, maxCount);

		return this.intern(
			RxKind.Interval,
			AsciiCharSet.empty,
			[child],
			low,
			maxCount,
			low === 0,
			maxLength);
	}

	/**
	 * The derivative of an expression after any character of a minterm.
	 *
	 * @param {Rx} expression The expression to differentiate.
	 * @param {AsciiCharSet} minterm A minterm of the expression's first sets. Every character in
	 * it must behave alike, which is what makes one derivative stand for the whole set.
	 * @returns {Rx} The derivative, which is the empty language where nothing follows.
	 */
	derivative(expression, minterm) {
		const key = `${expression.id}|${minterm.key()}`;
		const cached = this.derivatives.get(key);

		if (cached !== undefined) { return cached; }

		const result = this.computeDerivative(expression, minterm);

		this.derivatives.set(key, result);

		return result;
	}

	/**
	 * @param {Rx} expression The expression to differentiate.
	 * @param {AsciiCharSet} minterm The minterm.
	 * @returns {Rx} The derivative.
	 */
	computeDerivative(expression, minterm) {
		switch (expression.kind) {
			case RxKind.EmptySet:
			case RxKind.Epsilon:
				return this.emptySet;

			case RxKind.Chars:
				// The minterm is wholly inside the set or wholly outside it.
				return minterm.intersectsWith(expression.charSet) ? this.epsilon : this.emptySet;

			case RxKind.Concat: {
				const alternatives = [];

				for (let i = 0; i < expression.children.length; ++i) {
					const head = this.derivative(expression.children[i], minterm);

					if (head.kind !== RxKind.EmptySet) {
						alternatives.push(this.concat([head, ...expression.children.slice(i + 1)]));
					}

					// Only a part that can match nothing lets the character be consumed later on.
					if (!expression.children[i].isNullable) { break; }
				}

				return this.union(alternatives);
			}

			case RxKind.Union: {
				const alternatives = expression.children.map(
					child => this.derivative(child, minterm));

				return this.union(alternatives);
			}

			case RxKind.Interval: {
				const child = expression.children[0];
				const head = this.derivative(child, minterm);

				if (head.kind === RxKind.EmptySet) { return this.emptySet; }

				const minCount = expression.minCount === 0 ? 0 : expression.minCount - 1;

				return this.concatTwo(
					head,
					this.interval(child, minCount, expression.maxCount - 1));
			}

			default:
				throw new Error(`Unhandled kind ${expression.kind}.`);
		}
	}

	/**
	 * The identity of an expression is its shape and its operands, which are already interned and
	 * so are named by id.
	 *
	 * @param {string} kind One of {@link RxKind}.
	 * @param {AsciiCharSet} charSet The characters.
	 * @param {Rx[]} children The operands.
	 * @param {number} minCount The fewest repetitions.
	 * @param {number} maxCount The most repetitions.
	 * @param {boolean} isNullable Whether the language holds the empty string.
	 * @param {number} maxLength The longest string.
	 * @returns {Rx} The interned expression.
	 */
	intern(kind, charSet, children, minCount, maxCount, isNullable, maxLength) {
		const key = `${kind}|${charSet.key()}|${minCount}|${maxCount}|`
			+ children.map(child => child.id).join(',');

		const existing = this.interned.get(key);

		if (existing !== undefined) { return existing; }

		const created = new Rx(
			this.nextId++,
			kind,
			charSet,
			children,
			minCount,
			maxCount,
			isNullable,
			maxLength);

		this.interned.set(key, created);

		return created;
	}
}

/**
 * @param {number} left The first length.
 * @param {number} right The second length.
 * @returns {number} Their sum, saturated.
 */
function saturatingAdd(left, right) {
	const sum = left + right;

	return sum > MAX_LENGTH ? MAX_LENGTH : sum;
}

/**
 * @param {number} left The length.
 * @param {number} right The repetitions.
 * @returns {number} Their product, saturated.
 */
function saturatingMultiply(left, right) {
	if (left === 0 || right === 0) { return 0; }

	return left > MAX_LENGTH / right ? MAX_LENGTH : left * right;
}

File: src/cpp/src/rx.cpp
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

#include "array_view.hpp"
#include "rx.hpp"

#include "ascii_char_set.hpp"

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <memory>
#include <stdexcept>
#include <utility>
#include <vector>

namespace logmu::detail
{
	namespace
	{
		std::int64_t saturating_add(std::int64_t left, std::int64_t right) noexcept
		{
			return left > std::numeric_limits<std::int64_t>::max() - right
				? std::numeric_limits<std::int64_t>::max()
				: left + right;
		}

		std::int64_t saturating_multiply(std::int64_t left, std::int64_t right) noexcept
		{
			if (left == 0 || right == 0)
			{
				return 0;
			}

			return left > std::numeric_limits<std::int64_t>::max() / right
				? std::numeric_limits<std::int64_t>::max()
				: left * right;
		}
	}

	// rx

	const std::vector<ascii_char_set>& rx::get_first_sets() const
	{
		if (!this->first_sets_known)
		{
			this->collect_first_sets(this->first_sets);
			this->first_sets_known = true;
		}

		return this->first_sets;
	}

	void rx::collect_first_sets(std::vector<ascii_char_set>& sets) const
	{
		switch (this->kind)
		{
			case rx_kind::empty_set:
			case rx_kind::epsilon:
				return;

			case rx_kind::chars:
				sets.push_back(this->char_set);
				return;

			case rx_kind::concat:
				for (const rx* child : this->children)
				{
					child->collect_first_sets(sets);

					if (!child->is_nullable)
					{
						return;
					}
				}

				return;

			case rx_kind::union_:
				for (const rx* child : this->children)
				{
					child->collect_first_sets(sets);
				}

				return;

			case rx_kind::interval:
				this->children[0]->collect_first_sets(sets);
				return;
		}

		throw std::logic_error("Unhandled rx kind.");
	}

	// rx_factory

	rx_factory::rx_factory()
	{
		this->empty_set_node = this->intern(rx_kind::empty_set, ascii_char_set(), {}, 0, 0, false, 0);
		this->epsilon_node = this->intern(rx_kind::epsilon, ascii_char_set(), {}, 0, 0, true, 0);
	}

	const rx* rx_factory::chars(ascii_char_set set)
	{
		return set.is_empty()
			? this->empty_set_node
			: this->intern(rx_kind::chars, set, {}, 0, 0, false, 1);
	}

	const rx* rx_factory::concat(array_view<const rx*> parts)
	{
		std::vector<const rx*> flattened;

		for (const rx* part : parts)
		{
			if (part->kind == rx_kind::empty_set)
			{
				return this->empty_set_node;
			}

			if (part->kind == rx_kind::epsilon)
			{
				continue;
			}

			if (part->kind == rx_kind::concat)
			{
				flattened.insert(flattened.end(), part->children.begin(), part->children.end());
			}
			else
			{
				flattened.push_back(part);
			}
		}

		if (flattened.empty())
		{
			return this->epsilon_node;
		}

		if (flattened.size() == 1)
		{
			return flattened[0];
		}

		bool is_nullable = true;
		std::int64_t max_length = 0;

		for (const rx* part : flattened)
		{
			is_nullable = is_nullable && part->is_nullable;
			max_length = saturating_add(max_length, part->max_length);
		}

		return this->intern(rx_kind::concat, ascii_char_set(), std::move(flattened), 0, 0, is_nullable, max_length);
	}

	const rx* rx_factory::concat(const rx* first, const rx* second)
	{
		const rx* const parts[] = {first, second};

		return this->concat(parts);
	}

	const rx* rx_factory::union_(array_view<const rx*> alternatives)
	{
		std::vector<const rx*> flattened;

		for (const rx* alternative : alternatives)
		{
			if (alternative->kind == rx_kind::empty_set)
			{
				continue;
			}

			if (alternative->kind == rx_kind::union_)
			{
				flattened.insert(flattened.end(), alternative->children.begin(), alternative->children.end());
			}
			else
			{
				flattened.push_back(alternative);
			}
		}

		std::sort(flattened.begin(), flattened.end(), [](const rx* left, const rx* right) { return left->id < right->id; });

		std::vector<const rx*> distinct;
		distinct.reserve(flattened.size());

		for (const rx* alternative : flattened)
		{
			if (distinct.empty() || distinct.back() != alternative)
			{
				distinct.push_back(alternative);
			}
		}

		if (distinct.empty())
		{
			return this->empty_set_node;
		}

		if (distinct.size() == 1)
		{
			return distinct[0];
		}

		bool is_nullable = false;
		std::int64_t max_length = 0;

		for (const rx* alternative : distinct)
		{
			is_nullable = is_nullable || alternative->is_nullable;
			max_length = std::max(max_length, alternative->max_length);
		}

		return this->intern(rx_kind::union_, ascii_char_set(), std::move(distinct), 0, 0, is_nullable, max_length);
	}

	const rx* rx_factory::union_(const rx* first, const rx* second)
	{
		const rx* const alternatives[] = {first, second};

		return this->union_(alternatives);
	}

	const rx* rx_factory::interval(const rx* child, int min_count, int max_count)
	{
		if (max_count == 0)
		{
			return this->epsilon_node;
		}

		if (child->kind == rx_kind::epsilon)
		{
			return this->epsilon_node;
		}

		if (child->kind == rx_kind::empty_set)
		{
			return min_count == 0 ? this->epsilon_node : this->empty_set_node;
		}

		// Where the child accepts the empty string, so does every count above the minimum, and
		// x{m,n} and x{0,n} are the same language. Normalising here is what lets is_nullable be
		// read off the minimum alone.
		if (child->is_nullable)
		{
			min_count = 0;
		}

		if (min_count == 1 && max_count == 1)
		{
			return child;
		}

		const std::int64_t max_length = saturating_multiply(child->max_length, max_count);

		return this->intern(rx_kind::interval, ascii_char_set(), {child}, min_count, max_count, min_count == 0, max_length);
	}

	const rx* rx_factory::derivative(const rx* expression, ascii_char_set minterm)
	{
		const derivative_key key{expression->id, minterm};

		if (const auto found = this->derivatives.find(key); found != this->derivatives.end())
		{
			return found->second;
		}

		const rx* result = this->compute_derivative(expression, minterm);
		this->derivatives.emplace(key, result);

		return result;
	}

	const rx* rx_factory::compute_derivative(const rx* expression, ascii_char_set minterm)
	{
		switch (expression->kind)
		{
			case rx_kind::empty_set:
			case rx_kind::epsilon:
				return this->empty_set_node;

			case rx_kind::chars:
				// The minterm is wholly inside the set or wholly outside it.
				return minterm.intersects_with(expression->char_set) ? this->epsilon_node : this->empty_set_node;

			case rx_kind::concat:
			{
				std::vector<const rx*> alternatives;

				for (std::size_t i = 0; i < expression->children.size(); ++i)
				{
					const rx* head = this->derivative(expression->children[i], minterm);

					if (head->kind != rx_kind::empty_set)
					{
						std::vector<const rx*> parts;
						parts.reserve(expression->children.size() - i);
						parts.push_back(head);
						parts.insert(parts.end(), expression->children.begin() + static_cast<std::ptrdiff_t>(i) + 1, expression->children.end());

						alternatives.push_back(this->concat(parts));
					}

					// Only a part that can match nothing lets the character be consumed later on.
					if (!expression->children[i]->is_nullable)
					{
						break;
					}
				}

				return this->union_(alternatives);
			}

			case rx_kind::union_:
			{
				std::vector<const rx*> alternatives;
				alternatives.reserve(expression->children.size());

				for (const rx* child : expression->children)
				{
					alternatives.push_back(this->derivative(child, minterm));
				}

				return this->union_(alternatives);
			}

			case rx_kind::interval:
			{
				const rx* child = expression->children[0];
				const rx* head = this->derivative(child, minterm);

				if (head->kind == rx_kind::empty_set)
				{
					return this->empty_set_node;
				}

				const int min_count = expression->min_count == 0 ? 0 : expression->min_count - 1;

				return this->concat(head, this->interval(child, min_count, expression->max_count - 1));
			}
		}

		throw std::logic_error("Unhandled rx kind.");
	}

	const rx* rx_factory::intern(rx_kind kind, ascii_char_set char_set, std::vector<const rx*> children, int min_count, int max_count, bool is_nullable, std::int64_t max_length)
	{
		rx_key key{kind, char_set, std::move(children), min_count, max_count};

		if (const auto found = this->interned.find(key); found != this->interned.end())
		{
			return found->second;
		}

		auto created = std::make_unique<rx>(this->next_id++, kind, char_set, key.children, min_count, max_count, is_nullable, max_length);
		const rx* result = created.get();

		this->nodes.push_back(std::move(created));
		this->interned.emplace(std::move(key), result);

		return result;
	}

	std::size_t rx_factory::rx_key_hash::operator()(const rx_key& key) const noexcept
	{
		std::size_t accumulated = static_cast<std::size_t>(key.kind);
		accumulated = (accumulated * 31) + key.char_set.hash();
		accumulated = (accumulated * 31) + static_cast<std::size_t>(key.min_count);
		accumulated = (accumulated * 31) + static_cast<std::size_t>(key.max_count);

		for (const rx* child : key.children)
		{
			accumulated = (accumulated * 31) + static_cast<std::size_t>(child->id);
		}

		return accumulated;
	}

	std::size_t rx_factory::derivative_key_hash::operator()(const derivative_key& key) const noexcept
	{
		return (static_cast<std::size_t>(key.expression_id) * 31) ^ key.minterm.hash();
	}
}
